### PR TITLE
Comment explaining use of FlushProcessWriteBuffers in updating card table

### PR DESCRIPTION
**Pull request by acl33**
Hope this is right, and correctly reflects the various discussions we've had...

### CODE REVIEW

**Comment by acl33** [comment]
Note I've not mentioned Section 11.6 on self-modifying code (where an instruction is modified by a different linear address to which it is loaded from - i.e. we *might* need to flush even were the update not cross-thread, because the address written to (2 bytes after the start of the instruction) might not be in the same cache line as the start of the instruction....should I add that as well?

@mjp41 is this about what you were thinking?

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -7389,6 +7389,13 @@ int gc_heap::grow_brick_card_tables (uint8_t* start,
 
         seg_mapping_table = new_seg_mapping_table;
 
+        // The above write of the card table base address, is read by executing threads via the instruction cache.
+        // The update counts as cross-modifying code (section 8.1.3 of the Intel Software Development Manual),
+        // which requires all _executing_ threads to invaliadate their icache. FlushProcessWriteBuffers achieves
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: Typo: 'invalidate'

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -7389,6 +7389,13 @@ int gc_heap::grow_brick_card_tables (uint8_t* start,
 
         seg_mapping_table = new_seg_mapping_table;
 
+        // The above write of the card table base address, is read by executing threads via the instruction cache.
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: Nit: remove the comma after address

**Comment by adityamandaleeka** [comment]
LGTM modulo those minor comments.

**Comment by mjp41** [comment]
@acl33 should you add a Volume number?  It took as a while to find the right one?

Otherwise LGTM.

**Comment by mjp41** [comment]
Actually, I think the following comments should also be updated:
https://github.com/dotnet/coreclr/blob/2dfc74df1f9fe238eb2ec25c4764b8b65432b010/src/vm/gcenv.ee.cpp#L1281

File: src/gc/gc.cpp (hunk 2 above)
Comment (jkotas) [inline src/gc/gc.cpp]: I do not think that there is instruction cache involved here. `g_gc_card_table` is regular static, it won't be read via the instruction cache.

The comment about instruction cache should be next to https://github.com/dotnet/coreclr/blob/2dfc74df1f9fe238eb2ec25c4764b8b65432b010/src/vm/gcenv.ee.cpp#L1281 as @mjp41 pointed out. It is where we are patching the code inline.

**Comment by acl33** [comment]
Thanks, all - you are quite right, I got the wrong one, it's StompWriteBarrier in gcenv.ee.cpp that I meant - it even has the comment I meant to update. So, have moved there.

This leaves me unable to explain why we need the FPWB I originally commented in gc.cpp - at least on x86. If there's anything I can write then please shout ;-)...

**Comment by acl33** [comment]
@jkotas, @mjp41
Do you think the comment was also correct in its previous location, should I retain that? (This has changed totally since @adityamandaleeka 's first LGTM!!)

**Comment by jkotas** [comment]
> Do you think the comment was also correct in its previous location, should I retain that?

I do not think that this comment is not applicable in the original location.

**Pull request merged**